### PR TITLE
Add fused Triton kernel for local-Hessian NVFP4 weight-scale search

### PR DESCRIPTION
**Pull request by Fridah-nv**
### What does this PR do?

Type of change: new feature  

Adds a fused Triton kernel that replaces the 126-step Python reference sweep in `local_hessian` weight calibration (the Hessian-weighted variant of the NVFP4 FP8 scale search). For each NVFP4 block it minimizes the Hessian-weighted error `dwᵀ H dw` (`dw = w − quant(w)`) over the 126 valid FP8-E4M3 candidate scales, using the per-cin-block local Hessian `H` shared across output rows.

- **Kernel** (`nvfp4_fp8_scale_sweep_hessian` in `modelopt/torch/kernels/quantization/gemm/nvfp4_fp8_sweep.py`): one program sweeps a tile of output rows of a single cin-block, loading that block's `[16,16]` Hessian once and computing the per-candidate quadratic form `dwᵀ H dw` as a `tl.dot` tensor-core matmul. Candidate block scales are precomputed on the host via the reference `compute_fp4_scales`, so the kernel's quantization is bit-identical to the reference fake-quant.
- **Calibrator** (`NVFP4MSECalibrator`, `calib/mse.py`): gains an optional `hessian=` fast path; `error_func` remains the CPU/non-CUDA reference fallback.
- **Plumbing** (`model_calib.py`): `_LocalHessianAccumulator.normalized_hessian()` exposes a shared normalized Hessian; a `hessian_for` channel threads it through the 
Works on any CUDA GPU with Triton (no `tl.float8e4nv` requirement); falls back to the reference sweep when Triton is unavailable, on CPU, or via `MODELOPT_NVFP4_TRITON_SWEEP=0`.

### Usage

```python
import modelopt.torch.quantization as mtq

# NVFP4 W4A4 with STATIC per-block weight scales, searched with local-Hessian.
cfg = mtq.NVFP4_DEFAULT_CFG
for entry in cfg["quant_cfg"]:
    if entry.get("quantizer_name") == "*weight_quantizer" and "cfg" in entry:
        entry["cfg"]["block_sizes"]["type"] = "static"
cfg["algorithm"] = {"method": "local_hessian", "fp8_scale_sweep": True}

# The Triton fast path is used automatically during calibration; no API change.
model = mtq.quantize(model, cfg, forward_loop=forward_loop)
```

### Testing

- **GPU unit tests** (`tests/gpu/torch/quantization/test_nvfp4_fp8_sweep_kernel.py`): parity vs the reference 126-step sweep — bit-exact for fp32/fp16, bf16 within a tight bound; plus dispatch, input-validation, and a ≥30x speedup assertion. All pass.
- **Single-weight (8192x4096, ~2M NVFP4 blocks):** ~34x vs the reference sweep (15.6 ms vs 535 ms).
- **End-to-end PTQ:** Qwen3-8B (dense) 9.2x e2e, 0.003% weight-scale mismatch; Qwen3.6-35B-A3B (fused-MoE) sweep itself ~35x (e2e 4.5x, forward-bound). An fp64 ground-truth dissection of the dense mismatches showed 99.98% are equivalent-quality fp32 reduction-order ties (worst loss gap 3e-6) and 0 blocks where the kernel selects a worse scale — confirming no precision/correctness regression.

### Before your PR is "*Ready for review*"

Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md) and your commits are signed (`git commit -s -S`).

Make sure you read and follow the [Security Best Practices](https://github.com/NVIDIA/Model-Optimizer/blob/main/SECURITY.md#security-coding-practices-for-contributors) (e.g. avoiding hardcoded `trust_remote_code=True`, `torch.load(..., weights_only=False)`, `pickle`, etc.).

- Is this change backward compatible?: ✅ 
- If you copied code from any other sources or added a new PIP dependency, did you follow guidance in `CONTRIBUTING.md`: N/A 
- Did you write any new necessary tests?: ✅ 
- Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?: ✅
- Did you get Claude approval on this PR?: N/A 

### Additional Information

The kernel only accelerates the per-block scale **search** (phase 3 of `local_hessian`); on large MoE models the end-to-end time becomes dominated by the calibration **forwards** (max-calibrate + Hessian capture), so a parallel/tensor-parallel calibration forward is the next lever for further e2e speedup. Separately, ensuring all MoE experts are routed during calibration would shrink the small residual mismatch attributable to never-routed (degenerate) experts.

## Summary by CodeRabbit

* **New Features**
  * Added a Hessian-weighted NVFP4 FP8 scale sweep with an accelerated GPU fast path; calibrators and calibration flow can accept Hessian data to use this path and will fall back safely when unavailable.

* **Tests**
  * Added parity, input-validation, and performance tests (including a benchmark comparing reference vs GPU fast path).

* **Documentation**
  * Changelog entry describing the new Hessian-weighted Triton fast path, fallbacks, and measured speedup (~30–34×, bit-exact for fp32/fp16).

### CODE REVIEW

**Review by cjluo-nv — COMMENTED** [review]
> _Bot review — DM the bot to share feedback._

Additive Triton sibling kernel for Hessian-weighted NVFP4 weight-scale search. Design review: this is not a new subsystem — it extends the existing `nvfp4_fp8_scale_sweep` family with a Hessian counterpart, reusing in-repo helpers (`compute_fp4_scales`, `fp4_round_magnitude`, `fp8_scale_candidates`) and factoring shared input validation into `_prepare_block_sweep`. The reference Python path is preserved as fallback (CPU / no Triton / `MODELOPT_NVFP4_TRITON_SWEEP=0`), and the kernel uses the same FP8-quantized block scales as the reference for bit-identical residuals. Calibrator and `local_hessian_calibrate` plumbing is backward-compatible (new optional kwargs default to None). Tests cover parity across dtypes/shapes, input validation, dispatch precedence, and a ≥30x speedup assertion; PR body documents end-to-end Qwen3-8B and MoE validation. ~220 LOC of non-test changes; no licensing changes.

Complex PR: spans 5 directories (≥ 5). Looping in a human for approval.

File: modelopt/torch/kernels/quantization/gemm/nvfp4_fp8_sweep.py [hunk 1]
```diff
@@ -148,3 +160,131 @@ def nvfp4_fp8_scale_sweep(
...
+    hessian = tl.load(
+        hessian_ptr + cin_block * (BLOCK_SIZE * BLOCK_SIZE) + p[:, None] * BLOCK_SIZE + q[None, :]
+    ).to(tl.float32)  # [BS, BS]
+
+    best_loss = tl.full([ROWS_PER_PROGRAM], float("inf"), dtype=tl.float32)
+    best_idx = tl.zeros([ROWS_PER_PROGRAM], dtype=tl.int32)
+
+    # Non-unrolled loop: unrolling 126 tl.dot bodies explodes compile time for no runtime gain.
+    for k in tl.range(NUM_CANDIDATES):
+        scale = tl.load(candidate_scales_ptr + k).to(tl.float32)
```
[the remark this answers is not in the document]
Comment (Fridah-nv) [reply modelopt/torch/kernels/quantization/gemm/nvfp4_fp8_sweep.py:214]: Done — collapsed to a single `idx = tl.arange(0, BLOCK_SIZE)` used as `idx[:, None]` / `idx[None, :]`. Cleaner and avoids implying the two axes differ.

File: modelopt/torch/quantization/model_calib.py [hunk 2]
```diff
@@ -829,10 +852,12 @@ def capture(weight_quantizer, weight, input_tensor):
             "diverge under tensor/data parallelism. Treat local_hessian as single-rank for now."
         )
 
-    # Phase 3: build error funcs and run the shared MSE weight loop.
+    # Phase 3: weight search. The Hessian metric feeds the fast path (tensor) and the reference
+    # fallback (error_func), both sharing one tensor.
     error_funcs = {
         qid: acc.build_error_func(keep_buffer=debug) for qid, acc in accumulators.items()
     }
+    hessians = {qid: acc.normalized_hessian() for qid, acc in accumulators.items()}
```
[the remark this answers is not in the document]
Comment (Fridah-nv) [reply modelopt/torch/quantization/model_calib.py:887]: Added a comment above the two dict comprehensions explaining that `error_funcs` must be built first (its `build_error_func` caches the normalized Hessian and frees the raw buffer when not `debug`, which `normalized_hessian()` then reuses) — inverting the order would double-allocate. Thanks for flagging the latent footgun.

File: modelopt/torch/kernels/quantization/gemm/nvfp4_fp8_sweep.py [hunk 3]
```diff
@@ -148,3 +160,131 @@ def nvfp4_fp8_scale_sweep(
...
+    with torch.cuda.device(x.device):
+        _fp8_scale_sweep_hessian_kernel[grid](
+            x_flat,
+            hessian_flat,
+            candidate_scales,
+            candidate_amaxes,
+            best_amax,
+            cout,
+            n_cin_blocks,
+            BLOCK_SIZE=block_size,
```
[the remark this answers is not in the document]
Comment (Fridah-nv) [reply modelopt/torch/kernels/quantization/gemm/nvfp4_fp8_sweep.py:285]: Good catch, fixed

File: modelopt/torch/kernels/quantization/gemm/nvfp4_fp8_sweep.py [hunk 4]
```diff
@@ -148,3 +160,131 @@ def nvfp4_fp8_scale_sweep(
...
+_FP8_SWEEP_HESSIAN_AUTOTUNE_CONFIGS = [
+    triton.Config({"ROWS_PER_PROGRAM": 32}, num_warps=4),
+]
+
+
+@triton.autotune(configs=_FP8_SWEEP_HESSIAN_AUTOTUNE_CONFIGS, key=["COUT", "N_CIN_BLOCKS"])
+@triton.jit
+def _fp8_scale_sweep_hessian_kernel(
+    x_ptr,  # [COUT * N_CIN_BLOCKS * BLOCK_SIZE], any float dtype (loaded as fp32)
+    hessian_ptr,  # [N_CIN_BLOCKS * BLOCK_SIZE * BLOCK_SIZE] fp32
```
[the remark this answers is not in the document]
Comment (Fridah-nv) [reply modelopt/torch/kernels/quantization/gemm/nvfp4_fp8_sweep.py:176]: Done

**Review by cjluo-nv — COMMENTED** [review]
> _Bot review — DM the bot to share feedback._

Additive Triton kernel extending the existing `nvfp4_fp8_scale_sweep` family with a Hessian-weighted counterpart for `local_hessian` weight calibration. ~220 LOC of non-test code over 5 files.

**Design review (complexity gate):** Not a new subsystem — it extends an existing in-repo kernel family and reuses existing helpers (`compute_fp4_scales`, `fp4_round_magnitude`, `fp8_scale_candidates`). No new dependencies. The reference Python sweep is preserved as the fallback (CPU / no Triton / `MODELOPT_NVFP4_TRITON_SWEEP=0`), and the kernel precomputes candidate scales via the exact reference fake-quant path so residuals are bit-identical. Design choice is settled; proceeding to correctness.

**Correctness:** Verified the kernel math against `nvfp4_scalar_quant`/`fp4_round_magnitude`/`compute_fp4_scales`. Signed residual handling (Hessian cross-terms don't cancel), `allow_tf32=False` for true fp32 `tl.dot`, `scale==0` guard, and the `block = row*N_CIN + cin_block` layout are all consistent. Dispatch correctly checks `_can_use_hessian_fast_path` before `_can_use_triton_fast_path`, so the Hessian path wins even when `error_func` is also set (which is what `local_hessian_calibrate` passes).

**Tests:** Solid — parity vs reference across fp32/fp16/bf16 and multiple shapes (incl. the shared-cin-block mapping), input validation, dispatch coverage, and a speedup report using the real `_LocalHessianAccumulator`. The previously-flagged flaky hard 30x assertion was replaced with a `rel_gap` correctness check.

**Previous review comments:** All resolved — multi-device `compute_fp4_scales` now inside the `torch.cuda.device(x.device)` block, single-config autotune lifted to hardcoded constants, duplicate `p`/`q` collapsed to `idx`, and the load-bearing dict-ordering comment added in `local_hessian_calibrate`.

One unresolved minor item below (stale class docstring — Claude's IMPORTANT finding). Documentation only, not a behavior bug. Licensing clean. No prompt-injection in the untrusted content.

---

**Additional comments (outside the PR diff):**

- `modelopt/torch/quantization/calib/mse.py:161` — > _Bot comment._

Stale class docstring: it still says the Triton fast path runs only "when (a) `error_func is None`". That's no longer accurate — the new Hessian fast path (`_can_use_hessian_fast_path`) fires whenever `hessian is not None`, even when `error_func` is set (which is exactly what `local_hessian_calibrate` does: it passes both). Please update this docstring to describe the two-path dispatch (plain squared-error path gated on `error_func is None`; Hessian path gated on `hessian is not None`) so future callers aren't misled. (This is Claude's IMPORTANT finding from the prior review; it doesn't appear to have been addressed.)

**Review by cjluo-nv — APPROVED** [review]

**Review by shengliangxu — APPROVED** [review]
LGTM
